include/swift/Parse/ParsedRawSyntaxNode.h

```diff
@@ -0,0 +1,252 @@
+//===--- ParsedRawSyntaxNode.h - Parsed Raw Syntax Node ---------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_PARSE_PARSEDRAWSYNTAXNODE_H
+#define SWIFT_PARSE_PARSEDRAWSYNTAXNODE_H
+
+#include "swift/Basic/SourceLoc.h"
+#include "swift/Parse/Token.h"
+#include "swift/Syntax/SyntaxKind.h"
+#include "swift/Syntax/Trivia.h"
+#include <vector>
+
+namespace swift {
+
+typedef void *OpaqueSyntaxNode;
+
+/// Represents a raw syntax node formed by the parser.
+///
+/// It can be either 'recorded', in which case it encapsulates an
+/// \c OpaqueSyntaxNode that was returned from a \c SyntaxParseActions
+/// invocation, or 'deferred' which captures the data for a
+/// \c SyntaxParseActions invocation to occur later.
+///
+/// The deferred form is used for when the parser is backtracking and when
+/// there are instances that it's not clear what will be the final syntax node
+/// in the current parsing context.
+class ParsedRawSyntaxNode {
+  enum class DataKind: uint8_t {
+    Recorded,
```

Comment:
could you add some comments why we don't need `RecordedLayout` and `RecordedToken`?

Reply:
Done.